types/tls_test.go

```diff
@@ -33,16 +31,18 @@ func TestTLSVersion_Zero(t *testing.T) {
 
 func TestTLSVersion_ToJSON(t *testing.T) {
 	var tlsVersion TLSVersion
-	err := tlsVersion.UnmarshalJSON([]byte(`"foo"`))
-	require.Error(t, err)
-	require.Equal(t, tlsVersion, TLSVersionInvalid)
 
-	for str, version := range TLSVersions {
+	// Unmarshalling won't catch invalid version strings,
+	// must be checked in config or config entry validation
+	err := json.Unmarshal([]byte(`"foo"`), &tlsVersion)
```

Comment:
I wonder if we should include a custom `Marshal/Unmarshal` that validates the version.
Currently all codepaths are validated but if someone were to start using `types/tls` the type should provide guarantees of its enumerated states (or at least make it really hard to misuse).

Obviously someone could simply go ```TLSVersion("invalid version")``` and pass that around while breaking things but unmarshalling inputs into `TLSVersion` might lead to more subtle bugs.

Reply:
I had actually implemented this originally but removed it per @rboyer's suggestion to keep the validation pattern for config entries consistent - I ended up going down the rabbit hole of discovering that it's _very_ difficult to actually constrain the value of an enum-like type in Go 😞